### ISSUE TITLE
Issue with Vocabulary Size Divisibility and Worker Initialization

### ISSUE THREAD

**Issue by mahaddad** [vllm-project/vllm#1399]
Hello vLLM team and community,

First and foremost, thanks for all your hard work and dedication on this project.

**Issue Description:**

I've encountered an issue while trying to run inference for ehartford/WizardLM-Uncensored-Falcon-40b. The main problem seems to revolve around the vocabulary size (`65025`) not being divisible by the number of workers (`4`).

```python
python -m vllm.entrypoints.openai.api_server \
    --model ehartford/WizardLM-Uncensored-Falcon-40b
    --tensor-parallel-size 4
    --host 0.0.0.0 --port 8080 --trust-remote-code
```

Here's the traceback I received:

```plaintext
`
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/dist-packages/vllm/entrypoints/openai/api_server.py", line 616, in <module>
    engine = AsyncLLMEngine.from_engine_args(engine_args)
  File "/usr/local/lib/python3.8/dist-packages/vllm/engine/async_llm_engine.py", line 486, in from_engine_args
    engine = cls(engine_args.worker_use_ray,
  File "/usr/local/lib/python3.8/dist-packages/vllm/engine/async_llm_engine.py", line 270, in __init__
    self.engine = self._init_engine(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/vllm/engine/async_llm_engine.py", line 306, in _init_engine
    return engine_class(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/vllm/engine/llm_engine.py", line 106, in __init__
    self._init_workers_ray(placement_group)
  File "/usr/local/lib/python3.8/dist-packages/vllm/engine/llm_engine.py", line 179, in _init_workers_ray
    self._run_workers(
  File "/usr/local/lib/python3.8/dist-packages/vllm/engine/llm_engine.py", line 696, in _run_workers
    all_outputs = ray.get(all_outputs)
  File "/usr/local/lib/python3.8/dist-packages/ray/_private/auto_init_hook.py", line 24, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/ray/_private/worker.py", line 2547, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): ray::RayWorker.execute_method() (pid=2116, ip=192.168.2.4, actor_id=a63bb9515dc714ce4776550701000000, repr=<vllm.engine.ray_utils.RayWorker object at 0x7ca43b470160>)
  File "/usr/local/lib/python3.8/dist-packages/vllm/engine/ray_utils.py", line 32, in execute_method
    return executor(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/vllm/worker/worker.py", line 68, in init_model
    self.model = get_model(self.model_config)
  File "/usr/local/lib/python3.8/dist-packages/vllm/model_executor/model_loader.py", line 93, in get_model
    model = model_class(model_config.hf_config)
  File "/usr/local/lib/python3.8/dist-packages/vllm/model_executor/models/falcon.py", line 391, in __init__
    self.transformer = FalconModel(config)
  File "/usr/local/lib/python3.8/dist-packages/vllm/model_executor/models/falcon.py", line 349, in __init__
    self.word_embeddings = VocabParallelEmbedding(
  File "/usr/local/lib/python3.8/dist-packages/vllm/model_executor/parallel_utils/tensor_parallel/layers.py", line 107, in __init__
    VocabUtility.vocab_range_from_global_vocab_size(
  File "/usr/local/lib/python3.8/dist-packages/vllm/model_executor/parallel_utils/tensor_parallel/utils.py", line 67, in vocab_range_from_global_vocab_size
    per_partition_vocab_size = divide(global_vocab_size, world_size)
  File "/usr/local/lib/python3.8/dist-packages/vllm/model_executor/parallel_utils/tensor_parallel/utils.py", line 18, in divide
    ensure_divisibility(numerator, denominator)
  File "/usr/local/lib/python3.8/dist-packages/vllm/model_executor/parallel_utils/tensor_parallel/utils.py", line 10, in ensure_divisibility
    assert numerator % denominator == 0, "{} is not divisible by {}".format(
AssertionError: 65025 is not divisible by 4
`
```

**Model Configuration:**

Below is the model configuration I'm using:

```json
{
  "_name_or_path": "/workspace/WizardLM-Uncensored-falcon-40b",
  "alibi": false,
  "apply_residual_connection_post_layernorm": false,
  "architectures": [
    "RWForCausalLM"
  ],
  "attention_dropout": 0.0,
  "auto_map": {
    "AutoConfig": "configuration_RW.RWConfig",
    "AutoModelForCausalLM": "modelling_RW.RWForCausalLM"
  },
  "bias": false,
  "bos_token_id": 1,
  "eos_token_id": 2,
  "hidden_dropout": 0.0,
  "hidden_size": 8192,
  "initializer_range": 0.02,
  "layer_norm_epsilon": 1e-05,
  "model_type": "RefinedWeb",
  "n_head": 128,
  "n_head_kv": 8,
  "n_layer": 60,
  "parallel_attn": true,
  "torch_dtype": "float16",
  "transformers_version": "4.28.1",
  "use_cache": true,
  "vocab_size": 65025
}
```

I'm also seeing these warnings one of which may be relevant and the other which is likely not.

```plaintext
WARNING 10-17 10:32:40 config.py:383] The model's config.json does not contain any of the following keys to determine the original maximum length of the model: ['max_position_embeddings', 'n_positions', 'max_seq_len', 'max_sequence_length', 'max_seq_length', 'seq_len']. Assuming the model's maximum length is 2048.
```

```plaintext
2023-10-17 10:32:42,177 WARNING services.py:1889 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 67104768 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing '--shm-size=10.24gb' to 'docker run' (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.
```

**Potential Solutions / Queries:**

1. Would it be possible to provide some guidance on how to configure the number of workers or any other workaround to address this issue?
2. Are there any known limitations or configurations I should be aware of when working with models of this nature?

**Additional Context:**

- Using Python version: 3.8
- Installed package versions:
```plaintext
vllm==0.2.0
fschat==0.2.30
accelerate==0.23.0
typing_extensions==4.5.0
uninstalled transformer_engine to avoid a bug with the OpenAI API
```

Thanks for your support on this issue.

**Comment by HanNayeoniee** [comment]
Hi! I faced the same issue with you.
In my case, vocab_size is odd and I'm trying to inference with 2 gpus.
Have you fixed this? Thanks in advance!

**Comment by WoosukKwon** [comment]
Closed as this was fixed by #1622.

**Comment by VariableExp0rt** [comment]
@WoosukKwon first of all thank you for the project. Secondly, does the mentioned fix here https://github.com/vllm-project/vllm/pull/1622 include where the GPU number is odd? An example, we're deploying phi3 medium instruct 4B, and due to the total KV heads being `10`, we set the `tensor-parallel-size` so that `10 % tp_size == 0`. However, we now hit this issue because the [default padding](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/vocab_parallel_embedding.py#L201) is `64`, making the padded vocab size `32064`, which is not divisible by `5`. Are we going about this the right way? Is there a workaround? I believe it is the same circumstances as #4424.
